Fixing the logic issue in TransformTypos::TransformDesignatedInitExpr…

GS-GOAT opened the pull request:
… #126113

-Transforming Indices: For array designators, transform the index expression and update ExprChanged if it's modified.
-Correct Initializer Check: Compare the transformed initializer against the original to accurately track changes.
-Single Initializer Transformation: The initializer is processed once, not per designator, as each DesignatedInitExpr has one initializer.

Fixes #126113

zyn0217 reviewed (commented):
Thanks. Though I think we should just fix the problematic condition expression, and other changes seem a bit unnecessary since they are already clear enough.

GS-GOAT commented:
@zyn0217 Please review.

zyn0217 commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -13665,15 +13665,15 @@ TreeTransform<Derived>::TransformDesignatedInitExpr(DesignatedInitExpr *E) {
     if (D.isArrayDesignator()) {
-      ExprResult Index = getDerived().TransformExpr(E->getArrayIndex(D));
-      if (Index.isInvalid())
+      ExprResult NewIndex = getDerived().TransformExpr(E->getArrayIndex(D));
```
Can you avoid renaming the variable? Thanks

GS-GOAT committed:
Fixing the logic issue in TransformTypos::TransformDesignatedInitExpr llvm#126113

-Transforming Indices: For array designators, transform the index expression
and update ExprChanged if it's modified.
-Correct Initializer Check: Compare the transformed initializer against the
original to accurately track changes.
-Single Initializer Transformation: The initializer is processed once, not
per designator, as each DesignatedInitExpr has one initializer.

Fixes llvm#126113

GS-GOAT commented:
@zyn0217 Made the suggested changes.
